packages/abitype/src/utils.ts

```diff
@@ -84,100 +81,103 @@ type BitsTypeLookup = {
-   */
-  TAbiParameter['type'] extends `${infer Head}[${
-      | ''
-      | `${SolidityFixedArrayRange}`}]`
-  ? /**
-     * Then, infer in the opposite direction, using the known `Head` to infer the exact `Size` value.
-     *
-     * | Input        | Size |
-     * | ------------ | ---- |
-     * | `${Head}[]`  | `""` |
-     * | `${Head}[3]` | `3`  |
-     */
-    TAbiParameter['type'] extends `${Head}[${infer Size}]`
-    ? // Check if size is within range for fixed-length arrays, if so create a tuple.
-      // Otherwise, create an array. Tuples and arrays are created with `[${Size}]` popped off the end
-      // and passed back into the function to continue reduing down to the basic types found in Step 1.
-      Size extends keyof SolidityFixedArraySizeLookup
-      ? Tuple<
-          AbiParameterToPrimitiveType<
-            Merge<TAbiParameter, { type: Head }>,
-            TAbiParameterKind
-          >,
-          SolidityFixedArraySizeLookup[Size]
-        >
-      : readonly AbiParameterToPrimitiveType<
-          Merge<TAbiParameter, { type: Head }>,
-          TAbiParameterKind
-        >[]
-    : never
+  MaybeExtractArrayParameterType<TAbiParameter['type']> extends [
+      infer Head extends string,
+      infer Size,
+    ]
+  ? AbiArrayToPrimitiveType<TAbiParameter, TAbiParameterKind, Head, Size>
   : // 4. If type is not basic, tuple, or array, we don't know what the type is.
   // This can happen when a fixed-length array is out of range (`Size` doesn't exist in `SolidityFixedArraySizeLookup`),
   // the array has depth greater than `Config['ArrayMaxDepth']`, etc.
   ResolvedRegister['StrictAbiType'] extends true
-  ? TAbiParameter['type'] extends infer TAbiType extends string
-    ? Error<`Unknown type '${TAbiType}'.`>
-    : never
+  ? Error<`Unknown type '${TAbiParameter['type'] & string}'.`>
   : // 5. If we've gotten this far, let's check for errors in tuple components.
   // (Happens for recursive tuple typed data types.)
   TAbiParameter extends { components: Error<string> }
   ? TAbiParameter['components']
   : unknown
 
-// TODO: Speed up by returning immediately as soon as named parameter is found.
-type _HasUnnamedAbiParameter<TAbiParameters extends readonly AbiParameter[]> =
-  TAbiParameters extends readonly [
-    infer Head extends AbiParameter,
-    ...infer Tail extends readonly AbiParameter[],
-  ]
-    ? Head extends { name: string }
-      ? Head['name'] extends ''
-        ? true
-        : _HasUnnamedAbiParameter<Tail>
-      : true
-    : false
+type AbiBasicType = Exclude<AbiType, SolidityTuple | SolidityArray>
+
+type AbiComponentsToPrimitiveType<
+  Components extends readonly AbiParameter[],
+  TAbiParameterKind extends AbiParameterKind,
+> = Components extends readonly []
+  ? []
+  : // Compare the original set of names to a "validated"
+  // set where each name is coerced to a string and undefined|"" are excluded
+  Components[number]['name'] extends Exclude<
```

Comment:
This is a replacement for `_HasUnnamedAbiParameter`. It leverages how unions compare to one another to ensure that after the named parameters are extracted, it remains unchanged.

I wasn't sure how you wanted to deal with someone specifying a top-level `string`, but that could be disallowed as a special case.